src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokenDocumentationProvider.java

```diff
@@ -0,0 +1,118 @@
+import com.intellij.lang.documentation.AbstractDocumentationProvider;
+import com.intellij.openapi.progress.ProcessCanceledException;
+import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.platform.backend.documentation.DocumentationTarget;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiNamedElement;
+import com.intellij.util.ThreeState;
+import com.intellij.util.containers.ContainerUtil;
+import com.redhat.devtools.lsp4ij.features.documentation.LSPDocumentationTarget;
+import com.redhat.devtools.lsp4ij.features.documentation.LSPDocumentationTargetProvider;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nls;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * Documentation provider for {@link LSPSemanticTokensFileViewProvider} files that provides quick navigation info
+ * based on the LSP documentation for a semantic token.
+ */
+public class LSPSemanticTokenDocumentationProvider extends AbstractDocumentationProvider {
+
+    private static final LSPDocumentationTargetProvider LSP_DOCUMENTATION_TARGET_PROVIDER = new LSPDocumentationTargetProvider();
+
+    @Override
+    @Nullable
+    @Nls
+    public String getQuickNavigateInfo(PsiElement targetElement,
+                                       PsiElement sourceElement) {
+        String documentation = getDocumentation(sourceElement, targetElement);
+        return documentation != null ? documentation : super.getQuickNavigateInfo(targetElement, sourceElement);
+    }
+
+    /**
+     * Provides external access to documentation for other parts of LSP4IJ, e.g.,
+     * {@link com.redhat.devtools.lsp4ij.usages.LSPUsageElementDescriptionProvider}.
+     *
+     * @param element the element for which documentation is requested
+     * @return documentation for the element, or null if none could be found
+     */
+    @Nullable
+    @ApiStatus.Internal
+    public static String getDocumentation(@Nullable PsiElement element) {
+        return getDocumentation(element, element);
+    }
+
+    @Nullable
+    private static String getDocumentation(@Nullable PsiElement sourceElement,
+                                           @Nullable PsiElement targetElement) {
+        LSPSemanticTokensFileViewProvider semanticTokensFileViewProvider = LSPSemanticTokensFileViewProvider.getInstance(sourceElement);
+        if (semanticTokensFileViewProvider != null) {
+            int offset = sourceElement.getTextOffset();
+            ThreeState isIdentifier = semanticTokensFileViewProvider.isIdentifier(offset);
+            if ((isIdentifier == ThreeState.YES) ||
+                // If the token is definitely not an identifier and it's not whitespace, give the benefit of the doubt
+                ((isIdentifier == ThreeState.NO) && !semanticTokensFileViewProvider.isWhitespace(offset))) {
```

Comment:
This helps ensure that when the user is holding down **Ctrl/Cmd** and mousing around a document, we're not invoking the language server for hover doc unless there's a very strong chance that the offset is over an identifier.